[Merged by Bors] - sql: improve database schema handling

ivan4th opened the pull request:
## Motivation

Currently, the database schema is always built by running a series of migrations, even on a fresh database. There's no place in the source code to look for the full database schema in a single place. If schema drift happens for whatever reason (running dev version, a bug in a coded migration, etc.), it may go undetected causing issues at some later point in time.

Another problem is that `state.sql` and `local.sql` databases aren't being handled in a consistent manner, with `sql.Open` / `sql.InMemory` being used for the state database with its associated migrations, and `localsql.Open` / `localsql.InMemory` being derived from them. This has some unpleasant consequences, most importantly that it is hard to implement coded migrations as they'll have to be referred explicitly from `node/node.go` and possibly in other places.

`syncv2` will require some coded migration(s) for storing synctrees in the database, and ahead of these changes, it would be nice to have this kind of refactoring done in `go-spacemesh`.

## Description

When a new database is created, use schema script instead of running all the migrations. Also, check that there's no schema drift by diffing database schema against the schema script after running migrations.

Use consistent package/folder structure for `local.sql` and `state.sql` databases:
* `sql/statesql` and `sql/localsql` handle the state db and the local db, respectively, with `Schema`, `InMemory` and `Open` functions provided. Coded migrations can be added to these packages, too
* `sql/statesql/schema/schema.sql` and `sql/statesql/schema/schema.sql` contain current database schemas
* `sql/statesql/schema/migrations/*.sql` and `sql/statesql/schema/migrations/*.sql` contain database migrations

When a new database is created, the schema snapshot from `schema.sql` is used to initialize it. If the database already exists, its version is checked and if it's old, the migrations are run. In both cases, the schema is retrieved from the database afterwards and diffed against the schema snapshot, logging warnings about any detected differences (schema drift).

Fixes #6026.

## Test Plan

Verify on testnet and mainnet nodes

## TODO

- [X] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update [changelog](../CHANGELOG.md) as needed

ivan4th committed:
sql: improve database schema handling

Use consistent package/folder structure for local.sql and state.sql
databases.

When a new database is created, use schema script instead of running
all the migrations. Also, check that there's no schema drift by
diffing database schema against the schema script after running
migrations.

ivan4th committed:
sql: fixup: malsync and rewards

ivan4th committed:
sql: add migration / schema drift related tests

ivan4th committed:
sql: make schema drift table ignore regexp configurable

ivan4th committed:
sql: refactor localsql / statesql tests

ivan4th committed:
Update CHANGELOG.md

ivan4th committed:
config: fix db presets for db schema drift detection

poszu commented on sql/database.go:
```diff
@@ -93,48 +91,18 @@ func WithConnections(n int) Opt {
-func WithMigrations(migrations []Migration) Opt {
+// WithEnableMigrations enables or disables migrations on the database.
+// The migrations are enabled by default.
+func WithEnableMigrations(enable bool) Opt {
```
I think having 2 methods gives a nicer interface and avoids using a boolean arg:
```go
func WithMigrationsEnabled() Opt
func WithMigrationsDisabled() Opt
```

Wdyt?

ivan4th replied:
Well, probably makes sense, although `WithMigrationsEnabled` is likely also redundant as the migrations are enabled by default

ivan4th replied:
Fixed

poszu commented on sql/database.go:
```diff
@@ -172,12 +140,41 @@ func WithQueryCacheSizes(sizes map[QueryCacheKind]int) Opt {
+	}
+}
+
+func withForceFresh(fresh bool) Opt {
+	return func(c *conf) {
+		c.forceFresh = fresh
+	}
+}
```
As `forceFresh` is disabled by default, how about:

```suggestion
func withForceFresh() Opt {
	return func(c *conf) {
		c.forceFresh = true
	}
}
```

ivan4th replied:
ok will fix, thanks

ivan4th replied:
Fixed

poszu commented on sql/malsync/malsync.go:
```diff
@@ -8,6 +8,11 @@ import (
+	return timestamp, err
+}
+
+func getSyncState(db sql.Executor) (time.Time, bool, error) {
```
What's the meaning of the `bool` return value?

ivan4th replied:
It's to see if we need to create a new row (the table has just 1 row) or use an existing one
It would probably better to use `INSERT ... ON CONFLICT` ("upsert") but that would require another migration to add a primary key to the table
So I'll make return values named I think

fasmat replied:
There are two callers of `getSyncState`, one only is interested in the timestamp and one only is interested in the boolean. Why not make `getSyncState` two functions then? 🙂 

```go
func GetSyncState(db sql.Executor) (time.Time, error) {
	var timestamp time.Time
	rows, err := db.Exec("select timestamp from malfeasance_sync_state",
		nil, func(stmt *sql.Statement) bool {
			v := stmt.ColumnInt64(0)
			if v > 0 {
				timestamp = time.Unix(v, 0)
			}
			return true
		})
	if err != nil {
		return time.Time{}, fmt.Errorf("error getting malfeasance sync state: %w", err)
	}
	if rows != 1 {
		return time.Time{}, fmt.Errorf("expected malfeasance_sync_state to have 1 row but got %d rows", rows)
	}
	return timestamp, nil
}

func hasSynced(db sql.Executor) (bool, error) {
	rows, err := db.Exec("select 1 from malfeasance_sync_state", nil, nil)
	if err != nil {
		return false, fmt.Errorf("error getting malfeasance sync state: %w", err)
	}
	if rows > 1 {
		return false, fmt.Errorf("expected malfeasance_sync_state to have 1 row but got %d rows", rows)
	}
	return rows == 1, nil
}
```

fasmat replied:
> It would probably better to use INSERT ... ON CONFLICT ("upsert") but that would require another migration to add a primary key to the table

I'm actually for the suggested migration. Why not enforce on DB level that there can only be one timestamp in the DB instead of doing multiple checks in code to ensure there is only one? The migration should also be quick to apply or not? 🙂

ivan4th replied:
Added a migration and switched to using `INSERT ... ON CONFLICT`

poszu commented on sql/schema.go:
```diff
@@ -0,0 +1,202 @@
+	fmt.Fprintf(&sb, "PRAGMA user_version = %d;\n", version)
+	if _, err = db.Exec(
+		// Type is either 'index' or 'table', we want tables
+		// to go first. Also, we ignore _litestream tables
```
I think the comment is old. The `_litestream` tables can be filtered out with regex now, no?
```suggestion
		// to go first.
```

ivan4th replied:
Thanks, will fix

ivan4th replied:
Fixed

poszu commented on sql/schema.go:
```diff
@@ -0,0 +1,202 @@
+	if s.Script == actualScript {
+		return ""
+	}
+	diff := godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
+	if diff == "" {
+		return "<diff>"
+	}
+	return diff
```
I don't understand this part. Why not return the diff as it is? I'd assume that an empty `diff` means that schemas are equal.
```suggestion
	return godiffpatch.GeneratePatch(SchemaPath, s.Script, actualScript)
```

Also, why bother generating a diff at all? Is it for debugging purposes?

ivan4th replied:
> I'd assume that an empty `diff` means that schemas are equal.

Well I'd assume that too, but wanted to be on the safe side, in case if `GeneratePatch` doesn't work as expected.
Maybe it would be more appropriate to `panic` here, need to think about it.

> Also, why bother generating a diff at all? Is it for debugging purposes?

Yes, in case there are differences such as some index I've created by hand, an extra table column or a leftover from running an intermediate unreleased version it makes things much easier to see exactly what's the difference between the expected schema and the actual one.

fasmat replied:
> Maybe it would be more appropriate to panic here, need to think about it.

I don't think so? Does every call to `Diff` expect a difference between the actual script and what it is supposed to be? Also the log in `sql/database.go` is never called because this function doesn't ever return `""`.

fasmat replied:
> Yes, in case there are differences such as some index I've created by hand, an extra table column or a leftover from running an intermediate unreleased version it makes things much easier to see exactly what's the difference between the expected schema and the actual one.

In that case I would log the diff as `Debug` log in `sql.Open` and return an error like `errors.New("open DB: schema drift")`, which should exit the node gracefully if that happens during startup and cause a panic when using `InMemory`.

fasmat replied:
Also: we shouldn't introduce a dependency just for logging a diff of two schemas. We already have for instance `github.com/google/go-cmp` in our dependencies that does basically the same as `godiffpatch`

ivan4th replied:
> I don't think so? Does every call to `Diff` expect a difference between the actual script and what it is supposed to be? Also the log in `sql/database.go` is never called because this function doesn't ever return `""`.

There's a check above that diff that checks if the schema has changes
```
	if s.Script == actualScript {
		return ""
	}
```
So diff is expected to return a non-empty string here

ivan4th replied:
> In that case I would log the diff as `Debug` log in `sql.Open` and return an error like `errors.New("open DB: schema drift")`, which should exit the node gracefully if that happens during startup and cause a panic when using `InMemory`.

It is often very useful to allow schema drift while debugging go-spacemesh. For example, I might want to check if there's some performance improvement if I create some index, and that also constitutes an event of schema drift. We could of course have a config key for tolerating schema drift, but I'm not 100% sure I've anticipated all the situations which might cause adding extra stuff to the database, so I'm not sure if there will be no complaints about go-sm not working

ivan4th replied:
Switched to go-cmp

fasmat replied:
> It is often very useful to allow schema drift while debugging go-spacemesh. For example, I might want to check if there's some performance improvement if I create some index, and that also constitutes an event of schema drift. We could of course have a config key for tolerating schema drift, but I'm not 100% sure I've anticipated all the situations which might cause adding extra stuff to the database, so I'm not sure if there will be no complaints about go-sm not working

Alternatively the check that returns the error if the diff isn't empty can also be commented out during debugging 😉 I don't see a reason to allow such a thing and build a node that actively has migration issues to then deploy it?

fasmat replied:
If you think that it makes sense however to allow drifts in some cases I would go for the configuration option - if drift is allowed log the warning as now, if it is not return an error which will shut down the node gracefully.

ivan4th replied:
Schema drift is now fatal by default

poszu commented on sql/test/test.go:
```diff
@@ -0,0 +1,83 @@
+	Close() error
+}
+
+type DBFuncs[DB Database] struct {
+	Schema   func() (*sql.Schema, error)
+	Open     func(uri string, opts ...sql.Opt) (DB, error)
+	InMemory func(opts ...sql.Opt) DB
+}
```
Why not use an interface instead? It would require creating an implementation forwarding to the functions but it would seem less hacky.

```suggestion
type DBFuncs[DB Database] interface {
	Schema(*sql.Schema, error)
	Open(uri string, opts ...sql.Opt) (DB, error)
	InMemory(opts ...sql.Opt) DB
}
```

ivan4th replied:
Well, that'll cost some extra tens of lines, but probably should be more correct indeed, will fix

fasmat replied:
At the moment the tests say nothing. For instance which package is tested here and what happens in those tests?

```go
var fns = test.DBFuncs[*Database]{Schema: Schema, Open: Open, InMemory: InMemory}

func TestDatabase_1(t *testing.T) {
	test.ExecuteFirstTest(t, fns)
}

func TestDatabase_2(t *testing.T) {
	test.ExecuteSecondTest(t, fns)
}
```

fasmat replied:
When inlining this generic test into the actual packages it's only ~20 lines of code more but its much clearer what each test does:

```bash
$ git diff --stat 
 sql/localsql/localsql_test.go | 63 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 sql/statesql/statesql_test.go | 63 ++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-----
 sql/test/test.go              | 83 -----------------------------------------------------------------------------------
 3 files changed, 116 insertions(+), 93 deletions(-)
```

ivan4th replied:
It's a bit of tradeoff b/c while I agree that inlining the tests makes them more readable here, yet it also makes it more likely for them to drift apart over time as `localsql` / `statesql` pkgs are updated, e.g. to include some coded migrations. The `TestSchema` test is especially undesirable to be "forked" like this, as it provides means of updating the schema should it drift from what you get after running all the migrations on an empty database. Maybe I could keep the test bodies imported, but add some clarifying comments?

fasmat replied:
But with the same argument the `statesql` and `localsql` packages can drift apart. They are basically identical in their code - why not implement those in a generic way?

I know DRY is a paradigm, but in cases like here I think DRYing up the code is actually an anti-pattern - it makes the code harder to read and debug and if `statesql` and `localsql` are changed in the future the generic test most likely becomes more and more complex to accommodate for this instead of just having two different (but simpler) tests.

ivan4th replied:
Removed the "golden" test, replacing it with `make generate` / `make test-generate`, and inlined the idempotent schema test in the `localsql` and `statesql` packages

fasmat commented on go.mod:
```diff
@@ -38,6 +38,7 @@ require (
 	github.com/rs/cors v1.11.0
 	github.com/santhosh-tekuri/jsonschema/v5 v5.3.1
 	github.com/seehuhn/mt19937 v1.0.0
+	github.com/sourcegraph/go-diff-patch v0.0.0-20240223163233-798fd1e94a8e
```
In my opinion we shouldn't introduce a new dependency just to log a difference between two database schemas.

fasmat replied:
We already have a dependency on e.g. `github.com/google/go-cmp` which does a similar thing and we use in testing. Since this is also just used to log the difference in a human readable format we can use the same dependency for both 🙂

ivan4th replied:
Done

fasmat replied:
you probably still need to `go mod tidy` because the dependency is still in the `go.mod` and causes the linter to complain.

fasmat commented on sql/database.go:
```diff
@@ -30,6 +28,9 @@ var (
 	ErrTooNew = errors.New("database version is too new")
+	// ErrOld is returned when the database version differs from the expected one
+	// and migrations are disabled.
+	ErrOld = errors.New("old database version")
```
NIT: naming could be better

```suggestion
	// ErrOldSchema is returned when the database version differs from the expected one
	// and migrations are disabled.
	ErrOldSchema = errors.New("old database version")
```

ivan4th replied:
Done

fasmat commented on config/config.go:
```diff
@@ -117,6 +117,7 @@ type BaseConfig struct {
 	DatabaseSkipMigrations       []int                   `mapstructure:"db-skip-migrations"`
 	DatabaseQueryCache           bool                    `mapstructure:"db-query-cache"`
 	DatabaseQueryCacheSizes      DatabaseQueryCacheSizes `mapstructure:"db-query-cache-sizes"`
+	DatabaseIgnoreTableRx        string                  `mapstructure:"db-ignore-table-rx"`
```
Should this be configurable by an operator of the node?

ivan4th replied:
yes, e.g. `_litestream` tables are added by litestream which is an external program (SQLite replication engine)

fasmat replied:
I feel like this is a source of potential issues in the future. Other tools might add other tables but the config only allows to specify one regex.

Also I think having extra tables in the db can be considered a schema drift and this configuration option might unintentionally hide this 😕

ivan4th replied:
Schema drift is now fatal by default, ignore regexp removed

fasmat commented on sql/statesql/statesql.go:
```diff
@@ -0,0 +1,54 @@
+)
+
+//go:embed schema/schema.sql schema/migrations/*.sql
+var embedded embed.FS
```
If you embed the `schema.sql` file directly some code can be simplified:

```suggestion
//go:embed schema/schema.sql
var schema string

//go:embed schema/migrations/*.sql
var migrations embed.FS
```

ivan4th replied:
Done

fasmat commented on sql/statesql/statesql.go:
```diff
@@ -0,0 +1,54 @@
+	// NOTE: coded state migrations can be added here
+	// They can be a part of this statesql package
+	return sql.LoadSchema(embedded, migrations)
+}
```
For instance this becomes:

```suggestion
// Schema returns the schema for the state database.
func Schema() (*sql.Schema, error) {
	sqlMigrations, err := sql.LoadSQLMigrations(migrations)
	if err != nil {
		return nil, err
	}
	// NOTE: coded state migrations can be added here
	// They can be a part of this statesql package
	return &sql.Schema{Script: schema, Migrations: sqlMigrations}, nil
}
```

ivan4th replied:
Done

fasmat commented on sql/localsql/localsql.go:
```diff
@@ -0,0 +1,59 @@
+)
+
+//go:embed schema/schema.sql schema/migrations/*.sql
+var embedded embed.FS
```
See `statesql.go` - the same can be done here, by embedding `schema/schema.sql` directly the `sql.LoadSchema` function becomes obsolete.

ivan4th replied:
Swiched to string embedding as suggested.
The schema tests I slightly extended and kept under `sql/test`, as it's really undesirable for the `localsql` and `statesql` schema tests to diverge over time

fasmat replied:
We should continue the discussion in the other thread 🙂

fasmat commented on sql/schema.go:
```diff
@@ -0,0 +1,202 @@
+)
+
+// LoadSchema loads the schema embedded in the executable.
+func LoadSchema(fsys fs.FS, migrations []Migration) (*Schema, error) {
```
This function isn't actually needed (see my other comments).

ivan4th replied:
Removed

fasmat commented on sql/schema.go:
```diff
@@ -0,0 +1,202 @@
+}
+
+// WriteToFile writes the schema to the corresponding updated schema file.
+func (s *Schema) WriteToFile(basedir string) error {
+	path := filepath.Join(basedir, UpdatedSchemaPath)
+	if err := os.WriteFile(path, []byte(s.Script), 0o777); err != nil {
+		return fmt.Errorf("error writing schema file %s: %w", path, err)
+	}
+	return nil
```
This function seems to only exist for tests and there it is also not used besides writing it to a file and logging that it has been written? Example from the test:

```go
			loadedScript, err := sql.LoadDBSchemaScript(db, "")
			require.NoError(t, err)
			expSchema, err := Schema()
			require.NoError(t, err)
			diff := expSchema.Diff(loadedScript)
			if diff != "" {
				s := &sql.Schema{Script: loadedScript}
				require.NoError(t, s.WriteToFile("."))
				t.Logf("updated schema written to %s", sql.UpdatedSchemaPath)
			}
			require.Empty(t, diff, "schema diff")
```

That could just be:

```go
			loadedScript, err := sql.LoadDBSchemaScript(db, "")
			require.NoError(t, err)
			expSchema, err := Schema()
			require.NoError(t, err)
			require.Equal(t, expSchema.Script, loadedScript)
```

This already prints a diff if they are not the same and doesn't require `Schema.Diff` and `Schema.WriteToFile`

ivan4th replied:
`WriteFile` is needed so that the updated file can be used to replace schema file easily.
The expected workflow is this: you never edit schema file. Instead, you add migrations.
After you do that the the test fails and it writes the updated file, which you can rename to `schema.sql` if changes there are what you expect after migration. After you rename the updated schema file to `schema.sql` the test should pass (unless you have flaky migration that does different schema changes each time you run it)

ivan4th replied:
Tried doing just `require.Equal` (keeping `WriteFile`), but this prints out the whole script in the unreadable escaped form (`\n` instead of newlines) before printing the diff, which makes test output much more cluttered

fasmat replied:
I manually edited the `schema.sql` and executed the tests here is the output:

```
--- FAIL: TestSchema (0.04s)
    --- FAIL: TestSchema/idempotent_migration (0.02s)
        /workspaces/spacemesh/go-spacemesh/sql/statesql/logger.go:146: 2024-06-05T20:34:47.296Z	INFO	running migrations	{"uri": "file:/tmp/TestSchemaidempotent_migration785241580/001/test.db", "current version": 0, "target version": 18}
        /workspaces/spacemesh/go-spacemesh/sql/statesql/logger.go:146: 2024-06-05T20:34:47.310Z	WARN	database schema drift detected	{"uri": "file:/tmp/TestSchemaidempotent_migration785241580/001/test.db", "diff": "  (\n  \t\"\"\"\n  \t... // 9 identical lines\n  \t    PRIMARY KEY (address, layer_updated DESC)\n  \t);\n+ \tCREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);\n  \tCREATE TABLE activesets\n  \t(\n  \t... // 138 identical lines\n  \t\"\"\"\n  )\n"}
        /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:52: 
            	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:52
            	Error:      	Not equal: 
            	            	expected: 1
            	            	actual  : 2
            	Test:       	TestSchema/idempotent_migration
            	Messages:   	expected 1 log messages
    --- FAIL: TestSchema/schema (0.02s)
        --- FAIL: TestSchema/schema/no_migrations (0.01s)
            /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:102: 
                	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:102
                	Error:      	Not equal: 
                	            	expected: ***abbreviated***
                	            	actual  : ***abbreviated***
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -11,2 +11,3 @@
                	            	 );
                	            	+CREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);
                	            	 CREATE TABLE activesets
                	Test:       	TestSchema/schema/no_migrations
        --- FAIL: TestSchema/schema/force_migrations (0.01s)
            /workspaces/spacemesh/go-spacemesh/sql/statesql/test.go:102: 
                	Error Trace:	/workspaces/spacemesh/go-spacemesh/sql/test/test.go:102
                	Error:      	Not equal: 
                	            	expected: ***abbreviated***
                	            	actual  : ***abbreviated***
                	            	
                	            	Diff:
                	            	--- Expected
                	            	+++ Actual
                	            	@@ -11,2 +11,3 @@
                	            	 );
                	            	+CREATE INDEX accounts_by_layer_updated ON accounts (layer_updated);
                	            	 CREATE TABLE activesets
                	Test:       	TestSchema/schema/force_migrations
FAIL
```

Yes it prints the full values of the strings (here abbreviated) but also just the diff.

fasmat replied:
> The expected workflow is this: you never edit schema file. Instead, you add migrations.
After you do that the the test fails and it writes the updated file, which you can rename to schema.sql if changes there are what you expect after migration.

I'm not sure I like this workflow. We had such tests before where the test (or in this case here the shipped schema) needs to be updated by hand after adding a new migration. I'm not sure however what the alternative should be. In another project we had a dedicated tool that would generate up and down migrations for different DBs based on a DSL we wrote the migrations in and it would also generate the expected schema of the fully migrated DB when executed. Maybe the test is an OK substitute for this at the moment 🤔

ivan4th replied:
Switched to a workflow that uses `make generate`

fasmat commented on sql/schema.go:
```diff
@@ -0,0 +1,202 @@
+	})
+}
+
+// Migrate performs database migration. In case if migrations are disabled, the database
+// version is checked but no migrations are run, and if the database is too old and
+// migrations are disabled, an error is returned.
+func (s *Schema) Migrate(logger *zap.Logger, db *Database, vacuumState int, enable bool) error {
```
I'm usually not a fan of passing bool operators like this. `Migrate` should in my opinion not change its behavior if `config.enableMigrations` is set to `true` or `false`. The caller should check if running migrations is desired or not and either fail directly or call `Migrate` if needed, rather than putting that logic into the `Migrate` method.

ivan4th replied:
`Migrate` still checks db version in both cases, whether the migrations are enabled or disabled. In case if they're enabled, the old db version causes the migrations to be run, and if they're disabled, an error is returned. Maybe this method needs to be renamed... Splitting it in two may not be very pretty (version comparison logic not done in Schema? Or `enable` still passed to one of methods split out?)

fasmat replied:
It's about separation of concerns. In my opinion The caller of `Migrate` should check if a migration is necessary of not and if it is call `Migrate` instead of always calling `Migrate` and migrate determining if migration is necessary.

For me it becomes more logical the following way:

- check current version and determine target version
- if those are not equal and `config.enableMigrations` is false return an error
  - otherwise call migrate - its already a no-op if `before` and `after` are the same.

ivan4th replied:
Split `Migrate` into `CheckDBVersion` and `Migrate`

fasmat commented on sql/localsql/schema/schema.sql:
```diff
@@ -0,0 +1,82 @@
+PRAGMA user_version = 8;
```
How do we ensure that this schema is in line with the migrations? Is there a test that fails if I add a migration and forget to update this here?

ivan4th replied:
You're never expected to edit schema file by hand. You run `TestSchema` in `localsql` and `statesql` package. The test runs all the migrations starting from an empty db and captures the new schema, which it compares to the stored one. If there are schema differences, the test writes an updated schema file into `schema/schema.sql.updated`. If the diff shown by the test is acceptable, you do
```console
$ mv sql/schema/schema.sql{.updated,}
```
and then the tests passes. Basically, the so called "golden master" approach.

ivan4th replied:
Consequently, you can't have wrong `user_version` here. If you edit it by hand without adding migrations for the version number to match, you will have `TestSchema` fail and display the change in the diff.

fasmat replied:
Now thinking about this instead of doing this manually by executing a test and moving files around this is something that should happen on `make generate` in my opinion, since this is a generated file and just like go scale encoding and mocks should be updated on `make generate`. Then we also don't need the tests since we are already checking in our CI that `make generate` is a no-op or fail the build (even before tests and builds are executed). Maybe there is a quick and easy way to make this happen?

ivan4th replied:
`schema.sql` files are now handled by `make generate`, and consequently verified by `make test-generate` which is a part of the CI workflow

fasmat commented on sql/statesql/schema/schema.sql:
```diff
@@ -0,0 +1,151 @@
+PRAGMA user_version = 18;
```
How do we ensure that this schema is in line with the migrations? Is there a test that fails if I add a migration and forget to update this here?

ivan4th replied:
See above

ivan4th committed:
sql: fix review comments

ivan4th committed:
sql: add database schema handling docs

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
config: fix build errors

fasmat commented on README.md:
```diff
@@ -514,6 +514,86 @@ $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
+* `sql/statesql/schema/migrations` for `state.sql`
+* `sql/localsql/schema/migrations` for `local.sql`
+
+Additionally, some migrations ("coded migrations") can be implemented in Go code, in which case they reside in `.go` files located in `sql/statesql` and `sql/localsql` packages, respectively. It is worth noting that old coded migrations can be removed at some point, rendering database versions that are *too* old unusable with newer go-spacemesh versions.
```
Putting them into `sql/statesql` and `sql/localsql` packages will almost definitly cause cyclic dependency issues. For instance any coded migration involving ATXs will most likely need to import the `activation` package, but that package already has a dependency on both`sql/statesql` and `sql/localsql` packages...

EDIT: they are not removed - they are replaced by `.sql` files that do the SQL migrations (adding columns, tables, etc.) if necessary (otherwise are just empty migrations) but don't migrate data.

ivan4th replied:
[Last time we did it](https://github.com/spacemeshos/go-spacemesh/blob/37546bc2c93417e8dfdc24e5cc7c82d841c1f480/sql/migrations/state_0017_migration.go) we didn't have to import `activation`. In most cases `common/types` might suffice, and in the worst case a bit of code may be copied over to the migration from packages like `sql/atxs`, given that migration code has relatively short lifespan and is removed after the older versions become completely unsupported

fasmat replied:
I seem to remember we ran into that issue with previous coded migrations for the local db.

ivan4th replied:
Fixed with a bit of extra interfaces, see here https://github.com/spacemeshos/go-spacemesh/pull/6003#issuecomment-2164193831

fasmat commented on README.md:
````diff
@@ -514,6 +514,86 @@ $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
+logger.go:146: 2024-06-05T05:39:32.247+0400	WARN	database schema drift detected	{"uri": "file:/var/folders/r0/4mks2v4n5ysbntnf3xq6h_q80000gn/T/TestSchemaidempotent_migration3425594786/001/test.db", "diff": "  (\n  \t\"\"\"\n  \t... // 81 identical lines\n  \t    PRIMARY KEY (kind, epoch)\n  \t) WITHOUT ROWID;\n- \t\n- \t-- some change\n  \t\"\"\"\n  )\n"}
+```
+
+In this case, an empty line and `-- some change` was added to `schema.sql` by hand. The pretty-printed diff looks like this:
+```
+  (
+  	"""
+  	... // 81 identical lines
+  	    PRIMARY KEY (kind, epoch)
+  	) WITHOUT ROWID;
+-
+- 	-- some change
+  	"""
+  )
+```
+
+The possible reasons for schema drift can be the following:
+* running an unreleased version of go-spacemesh using your data folder. The unreleased version may contain migrations that may be changed before the release happens
+* manual changes in the database
+* external SQLite tooling used on the database that adds some tables, indices etc.
+
+In the latter case, it is possible to make go-spacemesh ignore certain objects (tables and indices) when checking for schema drift. For this, you can use `main.db-schema-ignore-rx` setting to set a regular expression that is used to ignore tables and indices in the database during schema drift checks. The setting defaults to `_litestream` to help with certain tooling.
+
+The schema changes in go-spacemesh code should be always done by means of adding migrations. After that, the schema tests in `sql/localsql` and `sql/statesql` will start failing. When the tests fail, they display the difference between the schema stored in `schema.sql` and the schema that is loaded from the database after running all the migrations.
+If the schema changes shown in the diff are expected, the schema file needs to be updated.
````
The `README.md` is primarily written for users - and this is already TMI for users. If the average user runs into a schema drift it's very unlikely they will be able to resolve that themselves.

ivan4th replied:
Moved these docs to `CONTRIBUTING.md`

fasmat commented on README.md:
````diff
@@ -514,6 +514,86 @@ $ grpcurl -plaintext 127.0.0.1:9093 spacemesh.v1.DebugService.NetworkInfo
+The schema changes in go-spacemesh code should be always done by means of adding migrations. After that, the schema tests in `sql/localsql` and `sql/statesql` will start failing. When the tests fail, they display the difference between the schema stored in `schema.sql` and the schema that is loaded from the database after running all the migrations.
+If the schema changes shown in the diff are expected, the schema file needs to be updated.
+
+```console
+$ # run the tests
+$ eval $(make print-test-env) go test ./sql/localsql ./sql/statesql
+...
+=== RUN   TestSchema/schema/force_migrations
+    test.go:106: updated schema written to schema/schema.sql.updated
+    test.go:108: 
+        	Error Trace:	/Users/ivan4th/work/spacemesh/go-spacemesh/sql/test/test.go:108
+        	Error:      	Should be empty, but was   (
+        	            	  	"""
+        	            	  	... // 81 identical lines
+        	            	  	    PRIMARY KEY (kind, epoch)
+        	            	  	) WITHOUT ROWID;
+        	            	- 	-- some change
+        	            	  	"""
+        	            	  )
+        	Test:       	TestSchema/schema/force_migrations
+        	Messages:   	schema diff
+FAIL
+FAIL    github.com/spacemeshos/go-spacemesh/sql/localsql        0.163s
+ok      github.com/spacemeshos/go-spacemesh/sql/statesql        0.286s
+FAIL
+$ git status
+...
+Untracked files:
+  (use "git add <file>..." to include in what will be committed)
+        sql/localsql/schema/schema.sql.updated
+
+$ # update the schema file
+$ mv sql/localsql/schema/schema.sql{.updated,}
+
+$ # rerun the tests
+$ eval $(make print-test-env) go test -count=1 ./sql/localsql ./sql/statesql
+ok      github.com/spacemeshos/go-spacemesh/sql/localsql        0.166s
+ok      github.com/spacemeshos/go-spacemesh/sql/statesql        0.293s
+```
````
No user will clone our repository and run tests if their node stops working due to a schema drift - I think this should be removed from the `README.md`. Also you are kinda leaking information about your system with putting or workspace path into the README 😉

ivan4th replied:
Updated schema handling docs and moved them to `CONTRIBUTING.md`

ivan4th committed:
Fix README.md

ivan4th committed:
Fix go.mod / go.sum

ivan4th committed:
sql: use go:generate for database schema files

ivan4th committed:
sql: make schema drift fatal by default

ivan4th committed:
sql: remove db-ignore-schema-rx config option

Also, add a migration that removes stray tables from quicksynced DBs

ivan4th committed:
sql: avoid cyclic dependencies in future coded migrations

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
activation: fix test

ivan4th committed:
sql: fix tests

ivan4th committed:
sql: fix query cache handling

ivan4th committed:
sql: update mocks

ivan4th committed:
sql: fix QueryCache related mocks

ivan4th committed:
node: don't print error twice upon failure

ivan4th committed:
sql: fix schema drift on Windows

ivan4th committed:
sql: update docs on schema handling

ivan4th commented:
Updated the PR according to https://github.com/spacemeshos/go-spacemesh/issues/6026#issuecomment-2159063376
Additional changes made:
* worked around circular dependencies in coded migrations by making `sql.Database` an interface with an internal implementation and introducing derived interfaces `sql.StateDatabase` and `sql.LocalDatabase`. This way, `sql/statesql` and `sql/localsql` packages are only imported by tests in most other packages and not the non-test code, and thus using most go-spacemesh packages in migrations located under `sql/statesql` and `sql/localsql` will not cause circular dependencies
* moved database schema handling docs from `README.md` to `CONTRIBUTING.md`
* fixed printing the error which caused the node startup to fail twice, so that docs can be prettier and still reflect the current codebase

ivan4th committed:
sql, malsync: fix handling of context cancelation

ivan4th committed:
sql: split Schema.Migrate() method

ivan4th committed:
sql: another fix for Windows newlines in the schema

ivan4th committed:
merge-nodes: fix test naming

ivan4th committed:
sql: close db on schema errors

fasmat commented on sql/database.go:
```diff
@@ -30,6 +28,9 @@ var (
+	ErrOldSchema = errors.New("old database version")
 )
 
 const (
```
A few lines below here is the `generate` statement for `mockgen` changing this line to

```go
//go:generate mockgen -typed -package=mocks -destination=./mocks/mocks.go github.com/spacemeshos/go-spacemesh/sql Executor
```

Will only generate a mock for the `Executor` interface (the only one that is actually mocked in at least one test). Saving > 1600 lines of generated code that is not used.

ivan4th replied:
Fixed, thanks

fasmat commented on datastore/store.go:
```diff
@@ -32,16 +32,9 @@ type VrfNonceKey struct {
-	WithTx(context.Context, func(*sql.Tx) error) error
-	QueryCache() sql.QueryCache
-}
-
```
With the removal of this interface there are no interfaces that can be mocked any more in this file. The `//go:generate` statement can be removed and the package `datastore/mocks` be deleted.

ivan4th replied:
Fixed

fasmat commented on sql/database.go:
```diff
@@ -498,6 +538,20 @@ func (tx *Tx) Exec(query string, encoder Encoder, decoder Decoder) (int, error)
 	return exec(tx.conn, query, encoder, decoder)
 }
 
+func fixError(err error) error {
```
NIT:

```suggestion
func mapSqliteError(err error) error {
```

ivan4th replied:
Fixed. The original intent was only to check for `SQLITE_INTERRUPT` in this function which was happening after context cancellation, and thus such cancellations had wrong error returned from SQLite funcs, but later I made the func a bit more general, forgetting to update its name

fasmat commented on sql/localsql/localsql_test.go:
```diff
@@ -0,0 +1,74 @@
+	})
+	require.NoError(t, err)
+
+	require.NoError(t, err)
```
This isn't needed 🙂

ivan4th replied:
Fixed

fasmat commented on sql/statesql/statesql_test.go:
```diff
@@ -0,0 +1,74 @@
+	})
+	require.NoError(t, err)
+
+	require.NoError(t, err)
```
This isn't needed 🙂

ivan4th replied:
Fixed

fasmat commented on sql/schemagen/main.go:
```diff
@@ -0,0 +1,48 @@
+var (
+	level  = zap.LevelFlag("level", zapcore.ErrorLevel, "set log verbosity level")
+	dbType = flag.String("dbtype", "state", "database type (state, local, default state)")
+	output = flag.String("output", "", "output file (defaults to stdin)")
```
```suggestion
	output = flag.String("output", "", "output file (defaults to stdout)")
```

ivan4th replied:
Fixed

ivan4th committed:
sql, datastore: remove unneeded mocks

ivan4th committed:
sql: fix naming

ivan4th committed:
sql: remove unneeded assertions from tests

ivan4th committed:
sql: schemagen: fix help

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th commented:
It might make sense to move the new section in `CONTRIBUTING.md` to `CODING.md` after #6045 is merged

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
Moved database schema handling docs to CODING.md

ivan4th commented:
Moved db schema docs to `CODING.md`

ivan4th committed:
api: fix database handling in the test

ivan4th committed:
sql: fix identities test

fasmat reviewed (commented):
This is OK from my side now; maybe @poszu also wants to take anoter look?

poszu commented on sql/database.go:
```diff
@@ -61,29 +62,27 @@ type Encoder func(*Statement)
+	logger            *zap.Logger
+	schema            *Schema
+	allowSchemaDrift  bool
+	ignoreSchemaDrift bool
```
It's confusing there are two fields that essentially mean the same thing. How about:
```suggestion
	allowSchemaDrift  bool // only useful with checkSchemaDrift = true
	checkSchemaDrift bool
```

ivan4th replied:
Fixed

poszu commented on sql/database.go:
```diff
@@ -277,13 +297,32 @@ func Version(uri string) (int, error) {
+type Transaction interface {
+	Executor
+	Commit() error
+	Release() error
```
Would it make sense to make `Release()` a mehod on the `Database` for symmetry (release the TX at the same place where it was acquired)?

ivan4th replied:
`Release()` is basically "rollback if not committed" in addition to releasing the connection, so it kind of also belongs to transaction just as well as `Commit`. Basically we have "Commit - Release" pair hear instead of "Commit - Rollback"

poszu commented on sql/database.go:
```diff
@@ -582,3 +636,15 @@ func LoadBlob(db Executor, cmd string, id []byte, blob *Blob) error {
+// StateDatabase is a Database used for Spacemesh state.
+type StateDatabase interface {
+	Database
+	IsStateDatabase() bool
```
NIT: It doesn't need to return anything if its purpose is to simply distinguish local vs state
```suggestion
	IsStateDatabase()
```

fasmat replied:
can it maybe even be a private method?

ivan4th replied:
Removed return value, but can't make these methods private as the interfaces are implemented in `sql/statesql` and `sql/localsql` packages and unexported interface methods can't be implemented in the packages other than the package where the interface is declared.

poszu commented on node/node.go:
```diff
@@ -1982,13 +1984,8 @@ func (app *App) setupDBs(ctx context.Context, lg log.Log) error {
-		return fmt.Errorf("load local migrations: %w", err)
-	}
 	localDB, err := localsql.Open("file:"+filepath.Join(dbPath, localDbFile),
 		sql.WithLogger(dbLog.Zap()),
-		sql.WithMigrations(migrations),
 		sql.WithConnections(app.Config.DatabaseConnections),
```
How about allowing schema drift for local DB too?

ivan4th replied:
Good point. Done

poszu commented on sql/malsync/malsync.go:
```diff
@@ -9,29 +9,33 @@ import (
-	return timestamp, nil
 }
 
 func updateSyncState(db sql.Executor, ts int64) error {
-	_, err := db.Exec("update malfeasance_sync_state set timestamp = ?1",
+	if _, err := db.Exec(
+		`insert into malfeasance_sync_state (id, timestamp) values(1, ?1)
+                   on conflict (id) do update set timestamp = ?1`,
```
Are these changes supposed to be part of this PR?

ivan4th replied:
The migration that added `malfeasance_sync_state` was a bit problematic for the new schema handling as it did insert a single row into the table. With the new schema handling, the fresh database is initialized from the schema file which is maintained by means of `go generate` and contains the schema captured from the SQLite database. The captured schema cannot contain any data, and thus it's no longer possible to depend on the fact that `malfeasance_sync_state` is not empty. Thus, I've added a dummy ID to this table to serve as a primary key, and instead of relying on the single row to be always present I use this "upsert" construct that creates the row when it doesn't exist, which will be the case with the fresh db.

ivan4th committed:
sql: simplify StateDatabase and LocalDatabase interfaces

ivan4th committed:
node: make it possibe to allow localsql schema drift

ivan4th commented:
bors merge

ivan4th commented:
bors merge

fasmat commented:
bors merge

fasmat commented:
bors merge

fasmat commented:
To me it looks like the problems were with the cluster not with the code changes. Those should have been fixed now, so lets try again 🙂 

bors merge

fasmat commented:
ok it seems like the problem is this PR 😕

ivan4th commented:
Tried to run the systests locally, didn't observe the problem so far.
There was k8s resource-related issue with systests yesterday, so will retry one more time

ivan4th commented:
bors merge

ivan4th commented:
The only test that failed in the last run, namely `TestPartition_50_50` passes in my env and likely will pass in the CI env if run in isolation. There appears to be a timing-dependent problem in systests (interaction between parallel tests) that this change somehow triggers, trying to look into that

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

fasmat commented:
bors merge

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th commented:
bors merge

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
sql: fix failing tests

ivan4th commented:
bors try

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th commented:
bors merge

ivan4th commented:
bors merge

ivan4th commented:
bors try

ivan4th commented:
bors merge

ivan4th commented:
windows unittest flake

ivan4th commented:
bors merge

ivan4th commented:
bors merge

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th commented:
Temporarily disabled `malfeasance_double_marry_test.go` via a build tag (circular dependency problem), will need to fix it before merging

ivan4th committed:
statesql: update schema

ivan4th committed:
tmp: fix lint errors (will need to revert)

ivan4th commented:
bors try

ivan4th commented:
Last `bors try` failed due to a flake in a unittest on Windows.
The systests have passed: https://github.com/spacemeshos/go-spacemesh/actions/runs/10368810152/job/28703211607
Doing another run to ensure stability.

ivan4th commented:
bors try

ivan4th committed:
Revert "tmp: fix lint errors (will need to revert)"

This reverts commit 1d8d3103606c40416720c52e3b525c4b50eb5db9.

ivan4th committed:
Move statesql migrations to a separate package to avoid cyclic deps

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
sql: fix schemagen

ivan4th commented:
bors try

ivan4th commented:
@fasmat @acud @poszu pls recheck this PR after the latest changes that were needed to resolve the conflicts

fasmat commented on sql/database.go:
```diff
@@ -511,6 +538,20 @@ func (tx *Tx) Exec(query string, encoder Encoder, decoder Decoder) (int, error)
 	return exec(tx.conn, query, encoder, decoder)
 }
 
+func mapSqliteError(err error) error {
+	code := sqlite.ErrCode(err)
+	if code == sqlite.SQLITE_CONSTRAINT_PRIMARYKEY || code == sqlite.SQLITE_CONSTRAINT_UNIQUE {
+		return ErrObjectExists
+	}
+	if code == sqlite.SQLITE_INTERRUPT {
+		// TODO: we probably should check if there was indeed a context
+		// that was canceled. But we're likely to replace crawshaw library
+		// in future so this part should be rewritten anyway
+		return context.Canceled
+	}
+	return err
+}
```
I don't think returning a `context.Canceled` with the assumption that if `SQLITE_INTERRUPT` happened a context must have been canceled somewhere is a good idea. From what I can see the returned error usually doesn't happen in a codepath that even is context aware.

Also I think its easier to parse what's happening here with a `switch`/`case` statement:

```suggestion
func mapSqliteError(err error) error {
	switch sqlite.ErrCode(err) {
	case sqlite.SQLITE_CONSTRAINT_PRIMARYKEY:
		return ErrObjectExists
	case sqlite.SQLITE_CONSTRAINT_UNIQUE:
		return ErrObjectExists
	default:
		return err
	}
}
```

ivan4th replied:
The code path that leads to this is the following:
1. When a transaction is initialized, a `context.Context` is passed to e.g. `sqliteDatabase.WithTx`
2. The context is used to acquire an SQLite connection from the pool
3. During this, the connection is set to be interrupted when the context is canceled: https://github.com/go-llsqlite/crawshaw/blob/v0.5.3/sqlitex/pool.go#L177
4. Without this extra check, this interruption during a long-running SQL query leads to a confusing error message that obscures the fact that the cause was context cancellation, and also breaks any error checks which look for `context.Canceled` errors

I actually did see this error on real nodes and thus I added extra check for this interruption.
The proper way to deal with it would be checking whether the actual context used to start the transaction is canceled in addition to checking the error code, but that would involve some extra modifications which I'd prefer to do in a separate PR.
So, I'm converting this code to a `switch`, but keeping the `sqlite.SQLITE_INTERRUPT`; also removing the comment about the need to replace `crawshaw`, as we decided against it in a separate discussion

ivan4th replied:
Issue created: #6273
There are further concerns mentioned in the issue about the fact that interrupting any writes to the DB will not rollback the transaction and will cause deadlocks later on.

fasmat replied:
Regarding the issue with a transaction not being rolled back on ctx cancellation: go 1.21 introduced [context.AfterFunc](https://pkg.go.dev/context#AfterFunc). Witin `WithTx` we could add `tx.Release` as an `AfterFunc` that gets executed when the context is cancelled and releases/rolls back the transaction. Wdyt?

ivan4th replied:
This will likely help, but it doesn't fully resolve the related write deadlock issue, and also there's still need to interrupt a long-running query which will always result in `sqlite.SQLITE_INTERRUPT`. In any case, I think it makes sense to do such changes in another PR, as this one is quite large and long-lasting already, on top of being in the need of constant resolution of merge conflicts

fasmat commented on sql/localsql/localsql.go:
```diff
@@ -0,0 +1,83 @@
+	db, err := Open("file::memory:?mode=memory", opts...)
+	if err != nil {
+		panic(err)
+	}
```
Would it make sense to simplify this to this?

```suggestion
	db := InMemory(opts...)
```

ivan4th replied:
Fixed

fasmat commented on sql/statesql/migrations/state_0021_migration_test.go:
```diff
@@ -13,44 +13,21 @@ import (
 )
 
-// Test that in-code migration results in the same schema as the .sql one.
-func Test0021Migration_CompatibleSchema(t *testing.T) {
```
I think we should keep this test, i.e. have a test available that opens a DB with the coded migration and one that opens it without and compares the schema of the two against each other.

This test right now is actually failing for me (without adjusting the `CREATE TABLE posts` statement in either the `.sql` or coded migration:

```go
// Test that in-code migration results in the same schema as the .sql one.
func Test0021Migration_CompatibleSchema(t *testing.T) {
	schema, err := statesql.Schema(New0021Migration(1000))
	require.NoError(t, err)

	db := statesql.InMemory(
		sql.WithLogger(zaptest.NewLogger(t)),
		sql.WithDatabaseSchema(schema),
		sql.WithForceMigrations(true),
	)
	schemasInCode, err := sql.LoadDBSchemaScript(db)
	require.NoError(t, err)
	require.NoError(t, db.Close())

	schema, err = statesql.Schema()
	require.NoError(t, err)

	db = statesql.InMemory(
		sql.WithLogger(zaptest.NewLogger(t)),
		sql.WithDatabaseSchema(schema),
		sql.WithForceMigrations(true),
	)
	require.NoError(t, err)

	schemasInFile, err := sql.LoadDBSchemaScript(db)
	require.NoError(t, err)
	require.NoError(t, db.Close())

	require.Equal(t, schemasInFile, schemasInCode)
}
```

wdyt?

ivan4th replied:
I actually added this test to the current version of this PR and it has passed.
The actual reason why it is not necessary is this:
- as of now, `schemagen` command which is used to generate `schema.sql` files for local and state databases imports and uses coded migrations
- `TestIdempotentMigration` in `sql/statesql` and `sql/localsql` uses SQL scripts ("placeholders") instead of coded migrations as it doesn't import them to avoid cyclic dependencies. If the coded migrations and SQL scripts result in different schemas, `TestIdempotentMigration` will fail, as it also uses `schema.sql` generated by `schemagen`

fasmat replied:
I think the reason as to why it failed was because of formatting of sql files on my end and yes `TestIdempotentMigration` would also have caught the issue. On that note see my other comment here: https://github.com/spacemeshos/go-spacemesh/pull/6003/files#r1723235468

fasmat commented on sql/statesql/statesql.go:
```diff
@@ -0,0 +1,81 @@
+	db, err := Open("file::memory:?mode=memory", opts...)
+	if err != nil {
+		panic(err)
+	}
```
Same as above:

```suggestion
	db := InMemory(opts...)
```

ivan4th replied:
Fixed

fasmat commented on syncer/malsync/syncer.go:
```diff
@@ -341,8 +340,15 @@ func (s *Syncer) downloadNodeIDs(ctx context.Context, initial bool, updates chan
+		return malsync.UpdateSyncState(tx, s.clock.Now())
+	}); err != nil {
+		if ctx.Err() != nil {
+			// FIXME: with crawshaw, canceling the context which has been used to get
+			// a connection from the pool may cause "database: no free connection" errors
```
Is this tracked as an issue? I only took a quick look at this issue and I'm not sure I fully understand the issue, or rather I believe we are not using `crawshaw` correctly in that aspect.

ivan4th replied:
Adding a mention of #6273 to the comment in the scope of which we need to check how well context cancellation is handled by crawshaw

fasmat commented:
Took me a bit but I made it through all changes, there is also still this comment by @poszu that is open: https://github.com/spacemeshos/go-spacemesh/pull/6003#discussion_r1654745181

ivan4th committed:
Merge branch 'develop' into feature/schema-snapshot

ivan4th committed:
Addressed comments

ivan4th commented:
> Took me a bit but I made it through all changes, there is also still this comment by @poszu that is open: [#6003 (comment)](https://github.com/spacemeshos/go-spacemesh/pull/6003#discussion_r1654745181)

Also fixed

fasmat commented on sql/schema.go:
```diff
@@ -0,0 +1,215 @@
+	skipMigration map[int]struct{}
+}
+
+// Diff diffs the database schema against the actual schema.
+// If there's no differences, it returns an empty string.
+func (s *Schema) Diff(actualScript string) string {
+	return cmp.Diff(s.Script, actualScript)
+}
```
Would it make sense to ignore whitespace?

```suggestion
// Diff diffs the database schema against the actual schema.
// If there's no differences, it returns an empty string.
func (s *Schema) Diff(actualScript string) string {
	// ignore whitespace differences
	opt := cmp.Comparer(func(x, y string) bool {
		return strings.Join(strings.Fields(x), "") == strings.Join(strings.Fields(y), "")
	})
	return cmp.Diff(s.Script, actualScript, opt)
}
```

ivan4th replied:
Fixed, thanks

ivan4th commented:
CI rerun due:
```
hare_test.go:653: eligibility can't be sent, waited: 10s, stacktraces:
```
Likely unrelated

ivan4th committed:
sql: ignore whitespace during schema drift checks

ivan4th commented:
bors merge
